### PR TITLE
Add rest highlevel explain API

### PR DESCRIPTION
**Pull request by PnPie**
Add explain API to rest high level client.

I create this PR firstly and I always have problem while parsing `GetReult` of `ExplainResponse` through XContent. I used `GetResult.toXContentEmbedded/fromXContentEmbedded` and it works well in `GetResultTests`, but in `ExplainResponseTests` seems it always fails with some arbitrary fields added in it. Anyone has any idea ?

Relates to #27205

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 1]
```diff
@@ -826,6 +855,14 @@ protected final ElasticsearchStatusException parseResponseException(ResponseExce
         }
     }
 
+    private <Resp> Resp processResponse(Resp resp, int statusCode) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: I see the problem here: the response doesn't print out the exists flag, which should be assigned based on the status code. I would not check instanceof though here, I think it's possible to provide a context to the fromXContent method as an additional argument (currently declared Void so it is always null), make that a Boolean . It is not very easy though to make that work based on the status in the client. As an alternative, we could change the output of the API and make it print out the flag as well, like many other API do. @nik9000 what do you think?

File: server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java [hunk 2]
```diff
@@ -129,4 +157,92 @@ public void writeTo(StreamOutput out) throws IOException {
...
+
+    private static final ConstructingObjectParser<ExplainResponse, Void> PARSER = new ConstructingObjectParser<>("explain", true,
+        arg -> new ExplainResponse((String) arg[0], (String) arg[1], (String) arg[2], (boolean) arg[3], (Explanation) arg[4],
+            (GetResult) arg[5]));
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), _INDEX);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), _TYPE);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), _ID);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), MATCHED);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java]: I think that you end up passing matched in as exists in the ExplainResponse constructor.

File: server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java [hunk 3]
```diff
@@ -21,19 +21,43 @@
...
+import java.util.Objects;
 
 import static org.elasticsearch.common.lucene.Lucene.readExplanation;
 import static org.elasticsearch.common.lucene.Lucene.writeExplanation;
 
 /**
  * Response containing the score explanation.
  */
-public class ExplainResponse extends ActionResponse {
+public class ExplainResponse extends ActionResponse implements ToXContentObject {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java]: you could make this StatusToXContentObject and have status method return the proper status based on the exists flag

File: server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java [hunk 4]
```diff
@@ -129,4 +157,92 @@ public void writeTo(StreamOutput out) throws IOException {
...
+                    return Explanation.match((float) arg[0], (String) arg[1], (Collection<Explanation>) arg[2]);
+                } else {
+                    return Explanation.noMatch((String) arg[1], (Collection<Explanation>) arg[2]);
+                }
+            });
+        explanationParser.declareFloat(ConstructingObjectParser.constructorArg(), VALUE);
+        explanationParser.declareString(ConstructingObjectParser.constructorArg(), DESCRIPTION);
+        explanationParser.declareObjectArray(ConstructingObjectParser.constructorArg(), explanationParser, DETAILS);
+        PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), explanationParser, EXPLANATION);
+        PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> GetResult.fromXContentEmbedded(p), GET);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java]: does this parse the explanation recursively like it's printed out in toXcontent?

File: server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java [hunk 5]
```diff
@@ -129,4 +157,92 @@ public void writeTo(StreamOutput out) throws IOException {
...
+    }
+
+    public static ExplainResponse fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/explain/ExplainResponse.java]: I think that you can remove this code from RestExplainAction and call this one instead by using RestStatusToXContentListener

File: server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import org.elasticsearch.test.AbstractStreamableXContentTestCase;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static org.hamcrest.Matchers.equalTo;
+
+public class ExplainResponseTests extends AbstractStreamableXContentTestCase<ExplainResponse> {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java]: I think it is expected that injecting random fields causes problems in some parts of the response, for instance the section where the source is returned. I guess that some path needs to be excluded using `getRandomFieldsExcludeFilter`

**Review by javanna — COMMENTED** [review]
I left some comments, thanks for your PR @PnPie !

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java (hunk 1 above)
Comment (nik9000) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: Could you make the `responseConverter` something like
```
response -> {
  ExplianResponse r = parseEntity(response.getEntity(), ExplainResponse::fromXContent);
  r.setExists(convertExistsResponse(response));
  return r;
}
```
?

Comment (nik9000) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: Ah! I see, you want the boolean as the context so you don't need to add `setExists`. You could do that like this:
```
response -> {
  CheckedFunction<XContentParser, Resp, IOException> entityParser = parser -> ExplainResponse.fromXContent(convertExistsResponse(response), parser);
  return parseEntity(response.getEntity(), entityParser);
}
```

You make ExplainResponse.fromXContent like this instead of how you have it.
```
    public static ExplainResponse fromXContent(boolean exists, XContentParser parser) {
        return PARSER.apply(parser, exists);
    }
```

I think that isn't *too* bad.

Comment (nik9000) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: And you'd have to change the parser declaration to:
```
private static final ConstructingObjectParser<ExplainResponse, Boolean> PARSER = new ConstructingObjectParser<>("explain", true, 
        (args, exists) -> new ExplainResponse((String) args[0], (String) args[1], (String) args[2], exists, (Explanation) args[3],
+            (GetResult) arg[4]));
```

Or something like that. I'm not 100% sure on that bit to be honest, but *somewhere* in the building function you'd use `exists`.
And then you can

**Review by PnPie — COMMENTED** [review]
Thanks for your review and comments @javanna @nik9000 ! I updated the PR.

File: server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            randomNonNegativeLong(),
+            true,
+            RandomObjects.randomSource(random()),
+            singletonMap(fieldName, new DocumentField(fieldName, values)));
+        return new ExplainResponse(index, type, id, exist, explanation, getResult);
+    }
+
+    @Override
+    protected Predicate<String> getRandomFieldsExcludeFilter() {
+        return field -> field.startsWith("fields") || field.startsWith("get");
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java]: I think that field.startsWith("get") could become field.startsWith("get._source") || field.startsWith("get.fields")

File: server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            "            \"field1\":[\n" +
+            "                \"value1\"\n" +
+            "            ]\n" +
+            "        }\n" +
+            "    }\n" +
+            "}").replaceAll("\\s+", "");
+        assertThat(expectedResponse, equalTo(generatedResponse));
+    }
+
+    private ExplainResponse createSimpleResponse() {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java]: given that we use this in a single place, can we remove this method and put the code where it's currently called?

File: server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        String id = "1";
+        boolean exist = true;
+        Explanation explanation = Explanation.match(1.0f, "description", Collections.emptySet());
+        GetResult getResult = new GetResult(null, null, null, -1, true, new BytesArray("{ \"field1\" : " +
+            "\"value1\", \"field2\":\"value2\"}"), singletonMap("field1", new DocumentField("field1",
+            singletonList("value1"))));
+        return new ExplainResponse(index, type, id, exist, explanation, getResult);
+    }
+
+    private Explanation randomExplanation(Explanation... explanations) {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java]: can this be static?

**Comment by javanna** [comment]
test this please

**Review by javanna — APPROVED** [review]
I left a couple of comments, this looks good though, thanks @PnPie !

File: server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java (hunk 7 above)
Comment (PnPie) [reply server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java]: Hi @javanna I change this line to `field.startsWith("get._source") || field.startsWith("get.fields")` but it still adds random items to "get.fields", it seems in this case it randomly inserts when field is "get" ?
Because when I do this `field.startsWith("get._source") || field.startsWith("get.fields") || field.equals("get")` then it works well.
So I keep this as `field.startsWith("get")` ?

Comment (javanna) [reply server/src/test/java/org/elasticsearch/action/explain/ExplainResponseTests.java]: good point, that is because GetResult prints out fields also at the `get` level, depending on whether they are metadata fields or not. So any random field at the `get` level will be parsed back as a field and printed out as part of `fields`. 

You can use `field.equals("get") || field.startsWith("get.fields") || field.startsWith("get._source")`

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 10]
```diff
@@ -593,6 +594,19 @@ static Request existsAlias(GetAliasesRequest getAliasesRequest) {
...
     }
 
+    static Request explain(ExplainRequest explainRequest) throws IOException {
+        Request request = new Request(HttpGet.METHOD_NAME,
+            endpoint(explainRequest.index(), explainRequest.type(), explainRequest.id(), "_explain"));
+
+        Params params = new Params(request);
+        params.withStoredFields(explainRequest.storedFields());
+        params.withFetchSourceContext(explainRequest.fetchSourceContext());
+        params.withRouting(explainRequest.routing());
```
[the remark this answers is not in the document]
Comment (javanna) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: Correct, we have removed support for parent in master. I will have to remember this when backporting to 6.x where we still support parent.

File: docs/java-rest/high-level/search/explain.asciidoc [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+An `ExplainRequest` expects an `index`, a `type` and an `id` to specify a certain document,
+and a query represented by `QueryBuilder` to run against it (the way of <<java-rest-high-query-builders, building queries>>).
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/SearchDocumentationIT.java[explain-request]
+--------------------------------------------------
+
+[[java-rest-high-explain-request-optional]]
+===== Optional arguments
```
Comment (javanna) [inline docs/java-rest/high-level/search/explain.asciidoc]: I think that this should be at the same level as request. I noticed some other pages where we do it this way but I prefer to have them at the same level.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (javanna) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: scratch that, I think this will be fine as-is in 6.x as well.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java [hunk 12]
```diff
@@ -835,6 +877,174 @@ public void testRenderSearchTemplate() throws IOException {
         assertToXContentEquivalent(expectedSource, actualSource, XContentType.JSON);
     }
 
+    public void testExplainSimple() throws IOException {
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java]: I would prefer not using Simple in a method name, although it's a test. Nothing is really simple :) Call it testExplain?

**Review by javanna — COMMENTED** [review]
I left a couple more comments and replied to your latest one. Thanks for doing this @PnPie I will merge it as soon as the comments are addressed.

**Comment by javanna** [comment]
test this please

File: client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java (hunk 12 above)
Comment (PnPie) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java]: Yeah definitly agree xD

**Comment by PnPie** [comment]
Thx @javanna I updated it.

**Comment by javanna** [comment]
retest this please

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SearchDocumentationIT.java [hunk 13]
```diff
@@ -1046,7 +1131,7 @@ private void indexSearchTestData() throws IOException {
         assertTrue(authorsResponse.isAcknowledged());
 
         CreateIndexRequest reviewersRequest = new CreateIndexRequest("contributors")
-            .mapping("doc", "user", "type=keyword");
+            .mapping("doc", "user", "type=keyword,store=true");
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SearchDocumentationIT.java]: was this change necessary?

Comment (PnPie) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SearchDocumentationIT.java]: I have tested the ["stored_fields" flag](https://github.com/elastic/elasticsearch/pull/31387/files#diff-da6eb90ef1049f45aefed2e1dc3d715cR866) in `testExplain()` and I add an assertion [here](https://github.com/elastic/elasticsearch/pull/31387/files#diff-da6eb90ef1049f45aefed2e1dc3d715cR897), if we don't "store" this field we will not get any values in `GetResult`'s `fields`, so this assertion will not work. Otherwise we remove this assertion ?

Comment (PnPie) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SearchDocumentationIT.java]: I have tested the ["stored_fields" flag](https://github.com/elastic/elasticsearch/pull/31387/files#diff-da6eb90ef1049f45aefed2e1dc3d715cR866) in `testExplain()` and I add an assertion [here](https://github.com/elastic/elasticsearch/pull/31387/files#diff-da6eb90ef1049f45aefed2e1dc3d715cR897), if we don't "store" this field we will not get any values in `GetResult`'s `fields`, so this assertion will not work. Otherwise we remove this assertion ?

Comment (javanna) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SearchDocumentationIT.java]: got it, it's fine, I got confused by the github diff, it seemed like this was part of testFieldCaps only, but it's shared between different methods. All good!

**Comment by javanna** [comment]
thanks @PnPie !

**Comment by nik9000** [comment]
@javanna I bumped into a failure locally after merging this that reproduced so I pushed ce1d0f5afdc505430e8a8bddb6545e5329ac8b1b to fix it. Could you bundle that when you backport this?

**Comment by javanna** [comment]
@nik9000 thanks for catching and fixing, I will integrate your fix when backporting.
